File: gui/DesignerV1_55/MultipleDemo/MultipleDemo.pas
program MultipleDemo;

Uses exec,intuition,gadtools,graphics,amiga,diskfont,
     workbench,utility,MultipleDemoPascal,defs;

Var
  done        : boolean;
  class       : long;
  code        : word;
  pgsel       : pGadget;
  imsg        : pintuimessage;
  dummy       : long;
  twindowlist : tlist;
  pwinnode    : pwindownode;
  pwinnode2   : pwindownode;
  mymsgport   : pmsgport;
  loop        : long;
  count       : long;
  
function openonewin : boolean;
begin
  openonewin:=false;
  pwinnode2:=allocmem(sizeof(twindownode),memf_any or memf_clear);
  if pwinnode2<>nil then
    begin
      if openwindowpwinnodepwin(mymsgport,pwinnode2) then
        begin
          openonewin:=true;
          pwinnode2^.pwin^.userdata:=pointer(pwinnode2);
          addtail(@twindowlist,pnode(pwinnode2));
        end
       else
        freemem(pwinnode,sizeof(twindownode));
    end;
  count:=0;
  pwinnode2:=pwindownode(twindowlist.lh_head);
  while (pwinnode2^.ln_succ<>nil) do
    begin
      inc(count);
      gt_setsinglegadgetattr(pwinnode2^.pwingads[commonwin_gad2],pwinnode2^.pwin,
                             gtnm_number,count);
      pwinnode2:=pwinnode2^.ln_succ;
    end;
end;

procedure closeonewin(pwn : pwindownode);
begin
  remove(pnode(pwn));
  closewindowpwinnodepwin(pwn);
  freemem(pwinnode,sizeof(twindownode));
  count:=0;
  pwinnode2:=pwindownode(twindowlist.lh_head);
  while (pwinnode2^.ln_succ<>nil) do
    begin
      inc(count);
      gt_setsinglegadgetattr(pwinnode2^.pwingads[commonwin_gad2],pwinnode2^.pwin,
                             gtnm_number,count);
      pwinnode2:=pwinnode2^.ln_succ;
    end;
end;

Procedure ProcessMenuIDCMPCommonMenu( MenuNumber : word);
Var
  ItemNumber : Word;
  Item       : pMenuItem;
Begin
  while (MenuNumber<>MENUNULL) do
    Begin
      Item:=ItemAddress( CommonMenu, MenuNumber);
      ItemNumber:=ITEMNUM(MenuNumber);
      MenuNumber:=MENUNUM(MenuNumber);
      Case MenuNumber of
        CommonMenu_Options :
          Case ItemNumber of
            CommonMenu_Options_Item0 :
              Begin
                pwinnode2:=pwindownode(remhead(@twindowlist));
                while (pwinnode2<>nil) do
                  begin
                    if twindowlist.lh_head^.ln_succ=nil then
                      begin
                        addhead(@twindowlist,pnode(pwinnode2));
                        pwinnode2:=nil;
                      end
                     else
                      begin
                        closeonewin(pwinnode2);
                        pwinnode2:=pwindownode(remhead(@twindowlist));
                      end;
                  end;
              end;
            CommonMenu_Options_Item2 :
              Begin
                pwinnode2:=pwindownode(remhead(@twindowlist));
                while (pwinnode2<>nil) do
                  begin
                    closeonewin(pwinnode2);
                    pwinnode2:=pwindownode(remhead(@twindowlist));
                  end;
                done:=true;
              end;
           end;
       end;
      MenuNumber:=Item^.NextSelect;
    end;
end;

Begin
  newlist(@twindowlist);
  done:=false;
  if OpenLibs then
	Begin
	  mymsgport:=createmsgport;
	  if mymsgport<>nil then
		begin
		  if openonewin then
		    while (not done) do
	   		  begin
			    dummy:=Wait(bitmask(mymsgport^.mp_SigBit));
			    imsg:=GT_GetIMsg(mymsgport);
			    while (imsg <>nil ) do
			  	  begin
				    class:=imsg^.Class;
		  		    code:=imsg^.Code;
				    pgsel:=pgadget(imsg^.IAddress);
				    pwinnode:=pointer(imsg^.idcmpwindow^.userdata);
				    GT_ReplyIMsg(imsg);
				    case class of
				      IDCMP_GADGETUP :
                        Begin
                          Case pgsel^.gadgetid of
                            CommonWin_Gad0 :
                              Begin
                                if not openonewin then
                                  writeln('Could not open window.');
                              end;
                            CommonWin_Gad1 :
                              Begin
                                for loop:=1 to 5 do
                                  if not openonewin then
                                    writeln('Could not open window.');
                              end;
                           end;
                        end;
                      IDCMP_CLOSEWINDOW :
                        begin
                          closeonewin(pwinnode);
                          if twindowlist.lh_head^.ln_succ=nil then
                            done:=true;
                        end;
                      IDCMP_MENUPICK :
                        ProcessMenuIDCMPCommonMenu( Code );
				     end;
				    imsg:=GT_GetIMsg(mymsgport);
				  end;
			  end;
		  if commonmenu<>nil then
		    freemenus(commonmenu);
		  deletemsgport(mymsgport);
		end
	   else
		writeln('Cannot make msg port.');
	  CloseLibs;
	end
   else
	writeln('Cannot open libraries.');
end.

File: gui/DesignerV1_55/MultipleDemo/MultipleDemo.c
/* Compile me to get full executable. */

#include "multipledemoc.c"

int done=0;
ULONG count;
UWORD Code;
struct Gadget *Gad;
struct IntuiMessage *imsg;
struct MsgPort *MyMsgPort;
struct WindowNode *WinNode;
struct WindowNode *WinNode2;
struct List WindowList;

int OpenNewWin(void)
{
struct WindowNode * WN;
WN=AllocMem(sizeof(struct WindowNode),MEMF_CLEAR);
if (WN !=NULL)
	{
	OpenWindowWinNodeWin( MyMsgPort, WN );
	if (WN->Win==NULL)
		{
		FreeMem(WN,sizeof(struct WindowNode));
		}
	else
		{
		AddTail(&WindowList,(struct Node *)WN);
		WN->Win->UserData=(void *)WN;
		count=0;
		WinNode2=(struct WindowNode *)WindowList.lh_Head;
		while (WinNode2->ln_Succ != NULL)
			{
			count++;
			GT_SetGadgetAttrs(WinNode2->WinGadgets[2],WinNode2->Win, NULL, GTNM_Number, count, TAG_DONE);
			WinNode2=WinNode2->ln_Succ;
			}
		return(0);
		}
	}
return(1);
}

int CloseOneWin( struct WindowNode * WN)
{
	Remove((struct Node *)WN);
	CloseWindowWinNodeWin(WN);
	FreeMem(WN,sizeof(struct WindowNode));
	count=0;
	WinNode2=(struct WindowNode *)WindowList.lh_Head;
	while (WinNode2->ln_Succ != NULL)
		{
		count++;
		GT_SetGadgetAttrs(WinNode2->WinGadgets[2],WinNode2->Win, NULL, GTNM_Number, count, TAG_DONE);
		WinNode2=WinNode2->ln_Succ;
		}
}

void ProcessMenuIDCMPCommonMenu( UWORD MenuNumber)
{
UWORD MenuNum;
UWORD ItemNumber;
struct MenuItem *item;
while ((MenuNumber != MENUNULL))
	{
	item = ItemAddress( CommonMenu, MenuNumber);
	MenuNum = MENUNUM(MenuNumber);
	ItemNumber = ITEMNUM(MenuNumber);
	switch ( MenuNum )
		{
		case CommonMenu_Options :
			switch ( ItemNumber )
				{
				case CommonMenu_Options_Item0 :
					WinNode=(struct WindowNode *)WindowList.lh_Head;
					while (WinNode->ln_Succ->ln_Succ != NULL)
						{
						CloseOneWin(WinNode);
						WinNode=(struct WindowNode *)WindowList.lh_Head;
						}
					break;
				case CommonMenu_Options_Item2 :
					WinNode=(struct WindowNode *)WindowList.lh_Head;
					while (WinNode->ln_Succ != NULL)
						{
						CloseOneWin(WinNode);
						WinNode=(struct WindowNode *)WindowList.lh_Head;
						}
					break;
				}
			break;
		}
	MenuNumber = item->NextSelect;
	}
}

int main(void)
{
ULONG class;
if (OpenLibs()==0)
	{
	NewList(&WindowList);
	if ((MyMsgPort=CreateMsgPort())!=0)
		{
		if (OpenNewWin()==0)
			while(done==0)
				{
				Wait(1L << MyMsgPort->mp_SigBit);
				imsg=GT_GetIMsg(MyMsgPort);
				while (imsg != NULL )
					{
					class=imsg->Class;
					Code=imsg->Code;
					Gad=(struct Gadget *)imsg->IAddress;
					WinNode = (struct WindowNode *)imsg->IDCMPWindow->UserData;
					GT_ReplyIMsg(imsg);
					switch ( class )
						{
						case IDCMP_GADGETUP :
							switch ( Gad->GadgetID ) 
								{
								case CommonWin_Gad0 :
									OpenNewWin();
									break;
								case CommonWin_Gad1 :
									OpenNewWin();
									OpenNewWin();
									OpenNewWin();
									OpenNewWin();
									OpenNewWin();
									break;
								}
							break;
						case IDCMP_CLOSEWINDOW :
							CloseOneWin(WinNode);
							if (WindowList.lh_Head->ln_Succ==NULL)
							  done=1;
							break;
						case IDCMP_MENUPICK :
							ProcessMenuIDCMPCommonMenu( Code );
							if (WindowList.lh_Head->ln_Succ==NULL)
							  done=1;
							break;
						}
					imsg=GT_GetIMsg(MyMsgPort);
					}
				}
		else
			printf("Cannot open window.\n");
		if (CommonMenu != NULL)
		  FreeMenus(CommonMenu);
		DeleteMsgPort(MyMsgPort);
		}
	else
		printf("Cannot make message Port.\n");
	CloseLibs();
	}
else
	printf("Cannot open libraries.\n");
}
